spring/boot-autoconfigure/src/main/java/com/linecorp/armeria/spring/AbstractServiceRegistrationBean.java

```diff
@@ -0,0 +1,122 @@
+/*
+ *  Copyright 2018 LINE Corporation
+ *
+ *  LINE Corporation licenses this file to you under the Apache License,
+ *  version 2.0 (the "License"); you may not use this file except in compliance
+ *  with the License. You may obtain a copy of the License at:
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ */
+package com.linecorp.armeria.spring;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+import javax.validation.constraints.NotNull;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.server.Service;
+
+/**
+ * An abstract bean with information for registering an annotated service object. It enables micrometer
+ * monitoring of the service automatically.
+ */
+class AbstractServiceRegistrationBean<T, U> {
+    /**
+     * The annotated service object to register.
+     */
+    @NotNull
+    private T service;
+
+    /**
+     * A service name to use in monitoring.
+     */
+    @NotNull
+    private String serviceName;
+
+    /**
+     * The decorators of the annotated service object.
+     */
+    @NotNull
+    private List<Function<Service<HttpRequest, HttpResponse>,
+            ? extends Service<HttpRequest, HttpResponse>>> decorators = new ArrayList<>();
+
+    /**
+     * Returns the annotated service object registered to this bean.
+     */
+    @NotNull
+    public T getService() {
+        return service;
+    }
+
+    /**
+     * Registers an annotated service object.
+     */
+    public U setService(@NotNull T service) {
+        this.service = service;
+        return unsafeCast(this);
+    }
+
+    /**
+     * Returns this service name to use in monitoring.
+     */
+    @NotNull
+    public String getServiceName() {
+        return serviceName;
+    }
+
+    /**
+     * Sets service name to use in monitoring.
+     */
+    public U setServiceName(
+            @NotNull String serviceName) {
+        this.serviceName = serviceName;
+        return unsafeCast(this);
+    }
+
+    /**
+     * Returns the decorators of the annotated service object.
+     */
+    @NotNull
+    public List<Function<Service<HttpRequest, HttpResponse>,
+            ? extends Service<HttpRequest, HttpResponse>>> getDecorators() {
+        return decorators;
+    }
+
+    /**
+     * Sets the decorator of the annotated service object. {@code decorators} are applied to {@code service} in
+     * order.
+     */
+    public U setDecorator(
```

Comment:
`setDecorators()`?

Reply:
Intentionally changed to `setDecorator` because if I understand correctly, if we hold method name `setDecorator`, it does not break API compatibility. How do you think?

Reply:
I think we still break ABI compatibility because the method signature of `setDecorator()` will change, i.e. a user has to re-compile his/her source code.

To do this right, we need to keep the old `setDecorator()` signature, and add a new varargs method.